ledger/src/bank_forks_utils.rs

```diff
@@ -222,77 +222,101 @@ fn bank_forks_from_snapshot(
-                full_snapshot_archive_info.slot(),
-                incremental_snapshot_archive_info
-                    .as_ref()
-                    .map(SnapshotArchiveInfoGetter::slot)
-                    .unwrap_or(0),
+    let bank = if will_startup_from_snapshot_archives {
+        // Given that we are going to boot from an archive, the append vecs held in the snapshot dirs for fast-boot should
+        // be released.  They will be released by the account_background_service anyway.  But in the case of the account_paths
+        // using memory-mounted file system, they are not released early enough to give space for the new append-vecs from
+        // the archives, causing the out-of-memory problem.  So, purge the snapshot dirs upfront before loading from the archive.
+        snapshot_utils::purge_old_bank_snapshots(&snapshot_config.bank_snapshots_dir, 0, None);
+
+        let (bank, _) = snapshot_utils::bank_from_snapshot_archives(
+            &account_paths,
+            &snapshot_config.bank_snapshots_dir,
+            &full_snapshot_archive_info,
+            incremental_snapshot_archive_info.as_ref(),
+            genesis_config,
+            &process_options.runtime_config,
+            process_options.debug_keys.clone(),
+            None,
+            process_options.account_indexes.clone(),
+            process_options.limit_load_slot_count_from_snapshot,
+            process_options.shrink_ratio,
+            process_options.accounts_db_test_hash_calculation,
+            process_options.accounts_db_skip_shrink,
+            process_options.verify_index,
+            process_options.accounts_db_config.clone(),
+            accounts_update_notifier,
+            exit,
+        )
+        .expect("load bank from snapshot archives");
+        bank
+    } else {
+        let Some(bank_snapshot) = latest_bank_snapshot else {
+            error!(
+                "There is no local state to startup from. Ensure --{} is *not* set to \"{}\" and restart.",
+                use_snapshot_archives_at_startup::cli::LONG_ARG,
+                UseSnapshotArchivesAtStartup::Never.to_string(),
             );
-            if bank.slot() < latest_snapshot_archive_slot {
-                warn!(
-                    "Starting up from local state at slot {}, which is *older* \
-                    than the latest snapshot archive at slot {}. If this is not \
-                    desired, change the --use-snapshot-archives-at-startup \
-                    CLI option to \"always\" and restart.",
-                    bank.slot(),
-                    latest_snapshot_archive_slot,
-                );
-            }
+            process::exit(1);
+        };
 
-            bank
+        // If a newer snapshot archive was downloaded, it is possible that its slot is
+        // higher than the local bank we will load.  Did the user intend for this?
+        if bank_snapshot.slot < latest_snapshot_archive_slot {
+            debug_assert_eq!(
```

Comment:
It feels to me this should be a regular `assert_eq!` since our intent is that this should not occur.

The typical use of debug asserts is for sanity in performance sensitive code, that we don't want to hit actual performance. I don't think that applies in this case, since it's only checked once at start up.